core/src/main/java/ai/timefold/solver/core/impl/score/director/ValueRangeManager.java

```diff
@@ -358,59 +365,255 @@ public <T> CountableValueRange<T> getFromSolution(ValueRangeDescriptor<Solution_
+        } else if (leftValueRange != null && (sorter == null || Objects.equals(leftSorter, sorter))) {
+            // Return the left value if there is no sorter or if the left sorter is the same as the provided one.
+            return leftValueRange;
+        } else if (rightValueRange != null && Objects.equals(rightSorter, sorter)) {
+            // Return the right value only if the right sorter is the same as the provided one.
+            return rightValueRange;
+        }
+        // If the left sorter is null, and the given sorter is not, we replace the left value with a sorted one.
+        if (leftSorter == null) {
+            if (!(leftValueRange instanceof SortableValueRange sortableValueRange)) {
+                throw new IllegalStateException(
+                        "Impossible state: value range (%s) on planning solution (%s) is not sortable."
+                                .formatted(valueRangeDescriptor, cachedWorkingSolution));
+            }
+            var sorterAdapter = SelectionSorterAdapter.of(cachedWorkingSolution, sorter);
+            var valueRange = (CountableValueRange<T>) sortableValueRange.sort(sorterAdapter);
+            valueRangeList[valueRangeDescriptor.getOrdinal()] =
+                    new ValueRangeItem<>(entity, valueRange, sorter, rightValueRange, rightSorter);
+            return valueRange;
+        } else if (rightValueRange == null) {
+            var valueRange = fetchValueRangeFromEntity(valueRangeDescriptor, entity, sorter);
+            valueRangeList[valueRangeDescriptor.getOrdinal()] =
+                    new ValueRangeItem<>(entity, leftValueRange, leftSorter, valueRange, sorter);
+            return valueRange;
+        } else {
+            throw new IllegalStateException(
+                    "Impossible state: the value range (%s) with sorter (%s) does not align with the existing ascending (%s) and descending (%s) sorters."
+                            .formatted(valueRangeDescriptor, sorter, leftSorter, rightSorter));
+        }
+    }
+
+    /**
+     * Search for an identical bitset linked to another entity.
+     * The goal is to deduplicate the value ranges stored in memory.
+     *
+     * @param valueRangeDescriptor the entity value range descriptor
+     * @param entity the entity
+     * @param valueRange the entity value range
+     * @return returns an existing entity with a matching value range, or returns null if it does not exist.
+     */
+    private @Nullable <T> Object findEntityBitSetMatch(ValueRangeDescriptor<Solution_> valueRangeDescriptor, Object entity,
+            int entitySize, CountableValueRange<T> valueRange) {
+        // We create a BitSet from the range values
+        // and check if another identical range already exists to prevent duplication
+        var valueIndexMap = getIndexMapFromSolution(valueRangeDescriptor);
+        var valueRangeBitSet = getBitSetValueRange(valueRange, valueIndexMap);
```

Comment:
I wonder... are there situation where we could avoid this?

We know the size of the collection; if the sizes differ, they are different collections. We also know the identity of the collection; if they are the same object, they are by definition duplicates and we don't need to compare bitsets.

Any way how we could use that information for our benefit?